File: runtime/fleet-mobile/modules/fleet-console-view/android/src/main/java/com/dotobokuri/fleet/mobile/TargetStore.kt
package com.dotobokuri.fleet.mobile

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.net.ServerSocket
import java.net.URI
import java.net.InetAddress
import java.security.SecureRandom

internal data class LoopbackIdentity(val host: String, val port: Int) {
  val authority: String get() = "$host:$port"
  val origin: String get() = "https://$authority"

  companion object {
    fun create(random: SecureRandom = SecureRandom()): LoopbackIdentity {
      val bytes = ByteArray(3).also(random::nextBytes)
      val host = "127.${1 + (bytes[0].toInt() and 0xff) % 254}.${1 + (bytes[1].toInt() and 0xff) % 254}.${1 + (bytes[2].toInt() and 0xff) % 254}"
      val port = ServerSocket(0, 1, InetAddress.getByName(host)).use { it.localPort }
      return LoopbackIdentity(host, port)
    }
  }
}

/**
 * Durable saved-console deck. A saved target stores where to go and what certificate to trust —
 * never a credential; pairing secrets stay in CookieManager under the remote origin. The deck is
 * keyed by remote origin: re-adding a console at the same address replaces its entry (label,
 * fingerprint, loopback identity) instead of growing a duplicate.
 */
internal class TargetStore(context: Context) {
  private val preferences = context.applicationContext.getSharedPreferences("fleet-mobile-target", Context.MODE_PRIVATE)

  fun active(): PersistedTarget? = state().let { current -> current.targets.firstOrNull { it.origin == current.active } }

  fun list(): List<PersistedTarget> = state().targets

  fun find(origin: String): PersistedTarget? = state().targets.firstOrNull { it.origin == origin }

  /** Commit synchronously so visible target, local origin, and durable metadata cannot diverge. */
  fun upsert(target: PersistedTarget): Boolean {
    if (!isValid(target)) return false
    val current = state()
    return write(State(target.origin, current.targets.filterNot { it.origin == target.origin } + target))
  }

  fun remove(origin: String): Boolean {
    val current = state()
    val remaining = current.targets.filterNot { it.origin == origin }
    if (remaining.size == current.targets.size) return false
    return write(State(current.active.takeIf { it != origin }, remaining))
  }

  fun identityFor(candidate: AccessTarget, live: PersistedTarget?, hasLiveGateway: Boolean): LoopbackIdentity {
    val stored = find(candidate.origin)
    val occupied = hasLiveGateway && live != null && stored != null && live.loopback == stored.loopback
    return selectIdentity(candidate, stored, occupied) { LoopbackIdentity.create() }
  }

  /** A resumed target reuses its saved identity unless the live gateway is already bound to it. */
  fun resumeIdentity(
    stored: PersistedTarget,
    live: PersistedTarget?,
    hasLiveGateway: Boolean,
    fresh: () -> LoopbackIdentity = { LoopbackIdentity.create() },
  ): LoopbackIdentity =
    if (hasLiveGateway && live != null && live.loopback == stored.loopback) fresh() else stored.loopback

  private data class State(val active: String?, val targets: List<PersistedTarget>)

  private fun state(): State {
    migrateLegacy()
    val raw = preferences.getString(KEY, null) ?: return State(null, emptyList())
    val parsed = try {
      readState(JSONObject(raw))
    } catch (_: Exception) {
      null
    }
    if (parsed == null) {
      preferences.edit().remove(KEY).commit()
      return State(null, emptyList())
    }
    return parsed
  }

  private fun readState(value: JSONObject): State {
    val entries = mutableListOf<PersistedTarget>()
    val array = value.getJSONArray("targets")
    for (index in 0 until array.length()) {
      val target = try { readTarget(array.getJSONObject(index)) } catch (_: Exception) { continue }
      if (isValid(target) && entries.none { it.origin == target.origin }) entries.add(target)
    }
    val active = value.optString("active").takeIf { name -> entries.any { it.origin == name } }
    return State(active, entries)
  }

  private fun readTarget(value: JSONObject): PersistedTarget = PersistedTarget(
    origin = value.getString("origin"),
    hostname = value.getString("hostname"),
    port = value.getInt("port"),
    label = value.getString("label"),
    fingerprint = value.getString("fingerprint"),
    loopback = LoopbackIdentity(value.getString("localHost"), value.getInt("localPort")),
  )

  private fun write(state: State): Boolean {
    val targets = JSONArray()
    state.targets.forEach { target ->
      targets.put(
        JSONObject()
          .put("origin", target.origin)
          .put("hostname", target.hostname)
          .put("port", target.port)
          .put("label", target.label)
          .put("fingerprint", target.fingerprint)
          .put("localHost", target.loopback.host)
          .put("localPort", target.loopback.port),
      )
    }
    val value = JSONObject().put("targets", targets)
    state.active?.let { value.put("active", it) }
    return preferences.edit().putString(KEY, value.toString()).commit()
  }

  /** The single-target v2 record becomes a one-entry deck with that target active. */
  private fun migrateLegacy() {
    val raw = preferences.getString(LEGACY_KEY, null) ?: return
    if (preferences.getString(KEY, null) == null) {
      val legacy = try {
        readTarget(JSONObject(raw))
      } catch (_: Exception) {
        null
      }
      if (legacy != null && isValid(legacy)) write(State(legacy.origin, listOf(legacy)))
    }
    preferences.edit().remove(LEGACY_KEY).commit()
  }

  private fun isValid(target: PersistedTarget): Boolean {
    val uri = try { URI(target.origin) } catch (_: Exception) { return false }
    val port = if (uri.port == -1) 443 else uri.port
    return uri.scheme == "https" && uri.rawUserInfo == null && uri.rawQuery == null && uri.rawFragment == null &&
      (uri.rawPath.isNullOrEmpty() || uri.rawPath == "/") &&
      uri.host.equals(target.hostname, ignoreCase = true) && port == target.port &&
      target.port in 1..65535 && target.fingerprint.matches(Regex("^[0-9A-F]{64}$")) &&
      target.label.isNotBlank() && target.label.length <= 48 &&
      target.loopback.host.matches(Regex("^127(?:\\.(?:0|[1-9]\\d?|1\\d\\d|2[0-4]\\d|25[0-5])){3}$")) &&
      target.loopback.host != "127.0.0.1" && target.loopback.port in 1..65535
  }

  companion object {
    private const val KEY = "targets-v3"
    private const val LEGACY_KEY = "target-v2"

    internal fun selectIdentity(
      candidate: AccessTarget,
      active: PersistedTarget?,
      hasLiveActive: Boolean,
      fresh: () -> LoopbackIdentity,
    ): LoopbackIdentity {
      val sameTarget = active != null && active.origin == candidate.origin && active.fingerprint == candidate.fingerprint
      return if (sameTarget && !hasLiveActive) active.loopback else fresh()
    }
  }
}

File: runtime/fleet-mobile/modules/fleet-console-view/ios/Core/TargetStore.swift
import Foundation

// TargetStore.kt의 이식. 저장하는 것은 "어디로 갈지 + 어떤 인증서를 믿을지"뿐이고 크리덴셜은
// 절대 저장하지 않는다. deck은 원격 origin을 키로 하며, 동기 커밋으로 화면/로컬오리진/영속
// 메타데이터가 어긋나지 않게 한다.
//
// iOS 분기: Android는 타깃마다 랜덤 127.x.y.z를 쓰지만 iOS 기기는 127.0.0.1 외 루프백을
// 바인딩할 수 없으므로(EADDRNOTAVAIL) 루프백 host는 항상 127.0.0.1이고 격리는 랜덤 포트가
// 담당한다. 그래서 isValid의 "127.0.0.1 금지"를 "127.0.0.1 강제"로 뒤집는다. Android의
// target-v2 → v3 마이그레이션은 iOS에 이력이 없으므로 구현하지 않는다(v3만).

// SharedPreferences 대신 쓰는 키-값 저장소 추상화. 테스트는 인메모리, 프로덕션은 UserDefaults.
public protocol KeyValueStore: AnyObject {
  func string(forKey key: String) -> String?
  func set(_ value: String?, forKey key: String)
}

public final class InMemoryKeyValueStore: KeyValueStore {
  private var storage: [String: String] = [:]
  public init() {}
  public func string(forKey key: String) -> String? { storage[key] }
  public func set(_ value: String?, forKey key: String) {
    if let value { storage[key] = value } else { storage.removeValue(forKey: key) }
  }
}

public final class UserDefaultsKeyValueStore: KeyValueStore {
  private let defaults: UserDefaults
  // Android의 SharedPreferences("fleet-mobile-target")에 대응하는 앱 프라이빗 스코프.
  public init(suiteName: String = "fleet-mobile-target") {
    defaults = UserDefaults(suiteName: suiteName) ?? .standard
  }
  public func string(forKey key: String) -> String? { defaults.string(forKey: key) }
  public func set(_ value: String?, forKey key: String) {
    if let value { defaults.set(value, forKey: key) } else { defaults.removeObject(forKey: key) }
  }
}

public final class TargetStore {
  private let store: KeyValueStore
  public init(store: KeyValueStore) { self.store = store }

  private static let key = "targets-v3"

  public func active() -> PersistedTarget? {
    let current = state()
    return current.targets.first { $0.origin == current.active }
  }

  public func list() -> [PersistedTarget] { state().targets }

  public func find(_ origin: String) -> PersistedTarget? { state().targets.first { $0.origin == origin } }

  /// 동기 커밋으로 화면/로컬오리진/영속 메타데이터가 어긋나지 않게 한다.
  @discardableResult
  public func upsert(_ target: PersistedTarget) -> Bool {
    guard Self.isValid(target) else { return false }
    let current = state()
    return write(State(active: target.origin, targets: current.targets.filter { $0.origin != target.origin } + [target]))
  }

  @discardableResult
  public func remove(_ origin: String) -> Bool {
    let current = state()
    let remaining = current.targets.filter { $0.origin != origin }
    if remaining.count == current.targets.count { return false }
    return write(State(active: current.active == origin ? nil : current.active, targets: remaining))
  }

  public func identityFor(_ candidate: AccessTarget, live: PersistedTarget?, hasLiveGateway: Bool, fresh: () -> LoopbackIdentity) -> LoopbackIdentity {
    let stored = find(candidate.origin)
    let occupied = hasLiveGateway && live != nil && stored != nil && live!.loopback == stored!.loopback
    return Self.selectIdentity(candidate, stored, occupied, fresh)
  }

  /// 재개된 타깃은 라이브 게이트웨이가 이미 그 identity를 잡고 있지 않는 한 저장된 identity를 재사용한다.
  public func resumeIdentity(_ stored: PersistedTarget, live: PersistedTarget?, hasLiveGateway: Bool, fresh: () -> LoopbackIdentity) -> LoopbackIdentity {
    if hasLiveGateway, let live, live.loopback == stored.loopback { return fresh() }
    return stored.loopback
  }

  public static func selectIdentity(_ candidate: AccessTarget, _ active: PersistedTarget?, _ hasLiveActive: Bool, _ fresh: () -> LoopbackIdentity) -> LoopbackIdentity {
    let sameTarget = active != nil && active!.origin == candidate.origin && active!.fingerprint == candidate.fingerprint
    return (sameTarget && !hasLiveActive) ? active!.loopback : fresh()
  }

  // MARK: - State

  private struct State { let active: String?; let targets: [PersistedTarget] }

  private func state() -> State {
    guard let raw = store.string(forKey: Self.key) else { return State(active: nil, targets: []) }
    guard let parsed = Self.readState(raw) else {
      store.set(nil, forKey: Self.key) // 손상 데이터는 삭제.
      return State(active: nil, targets: [])
    }
    return parsed
  }

  private static func readState(_ raw: String) -> State? {
    guard
      let data = raw.data(using: .utf8),
      let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
      let array = object["targets"] as? [[String: Any]]
    else { return nil }
    var entries: [PersistedTarget] = []
    for item in array {
      guard let target = readTarget(item), isValid(target), !entries.contains(where: { $0.origin == target.origin }) else { continue }
      entries.append(target)
    }
    let activeName = object["active"] as? String
    let active = (activeName != nil && entries.contains { $0.origin == activeName }) ? activeName : nil
    return State(active: active, targets: entries)
  }

  private static func readTarget(_ value: [String: Any]) -> PersistedTarget? {
    guard
      let origin = value["origin"] as? String,
      let hostname = value["hostname"] as? String,
      let port = intOf(value["port"]),
      let label = value["label"] as? String,
      let fingerprint = value["fingerprint"] as? String,
      let localHost = value["localHost"] as? String,
      let localPort = intOf(value["localPort"])
    else { return nil }
    return PersistedTarget(
      origin: origin, hostname: hostname, port: port, label: label, fingerprint: fingerprint,
      loopback: LoopbackIdentity(host: localHost, port: localPort))
  }

  @discardableResult
  private func write(_ state: State) -> Bool {
    let targets: [[String: Any]] = state.targets.map { t in
      [
        "origin": t.origin, "hostname": t.hostname, "port": t.port, "label": t.label,
        "fingerprint": t.fingerprint, "localHost": t.loopback.host, "localPort": t.loopback.port,
      ]
    }
    var object: [String: Any] = ["targets": targets]
    if let active = state.active { object["active"] = active }
    guard
      let data = try? JSONSerialization.data(withJSONObject: object),
      let json = String(data: data, encoding: .utf8)
    else { return false }
    store.set(json, forKey: Self.key)
    return true
  }

  static func isValid(_ target: PersistedTarget) -> Bool {
    guard let uri = LocationUri.parse(target.origin) else { return false }
    let port = uri.port == -1 ? 443 : uri.port
    let pathOk = uri.rawPath.isEmpty || uri.rawPath == "/"
    let hostOk = uri.host?.caseInsensitiveCompare(target.hostname) == .orderedSame
    return uri.scheme == "https" && uri.rawUserInfo == nil && uri.rawQuery == nil && uri.rawFragment == nil
      && pathOk && hostOk && port == target.port && (1...65535).contains(target.port)
      && matchesFull("^[0-9A-F]{64}$", target.fingerprint)
      && !target.label.trimmingCharacters(in: .whitespaces).isEmpty && target.label.utf16.count <= 48
      // iOS 강제: 루프백 host는 127.0.0.1(다른 루프백은 기기에서 바인딩 불가).
      && target.loopback.host == "127.0.0.1" && (1...65535).contains(target.loopback.port)
  }

  private static func intOf(_ value: Any?) -> Int? {
    if let n = value as? NSNumber, CFGetTypeID(n) != CFBooleanGetTypeID() { return n.intValue }
    return nil
  }

  private static func matchesFull(_ pattern: String, _ input: String) -> Bool {
    guard let regex = try? NSRegularExpression(pattern: pattern) else { return false }
    let range = NSRange(input.startIndex..<input.endIndex, in: input)
    guard let m = regex.firstMatch(in: input, options: [], range: range) else { return false }
    return m.range == range
  }
}
